File: src/Basescape/StoresState.cpp
/*
 * Copyright 2010-2016 OpenXcom Developers.
 *
 * This file is part of OpenXcom.
 *
 * OpenXcom is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * OpenXcom is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with OpenXcom.  If not, see <http://www.gnu.org/licenses/>.
 */
#include "StoresState.h"
#include <sstream>
#include <iomanip>
#include "../Engine/Game.h"
#include "../Mod/Mod.h"
#include "../Engine/LocalizedText.h"
#include "../Engine/Options.h"
#include "../Interface/TextButton.h"
#include "../Interface/Window.h"
#include "../Interface/Text.h"
#include "../Interface/TextList.h"
#include "../Savegame/Base.h"
#include "../Mod/RuleItem.h"
#include "../Savegame/ItemContainer.h"


namespace OpenXcom
{


struct compareSpace
{

	bool operator()(const StoredItem &a, const StoredItem &b) const
	{
		if (a.storageSize == b.storageSize)
		{
			return Unicode::naturalCompare(a.displayName, b.displayName);
		}
		else
		{
			return (a.storageSize > b.storageSize);
		}
	}
};


/**
 * Initializes all the elements in the Stores window.
 * @param game Pointer to the core game.
 * @param base Pointer to the base to get info from.
 */
StoresState::StoresState(Base *base) : _base(base)
{
	// Create objects
	_window = new Window(this, 320, 200, 0, 0);
	_btnOk = new TextButton(300, 16, 10, 176);
	_txtTitle = new Text(310, 17, 5, 8);
	_txtItem = new Text(162, 9, 10, 32);
	_txtQuantity = new Text(57, 9, 172, 32);
	_txtSpaceUsed = new Text(57, 9, 229, 32);
	_lstStores = new TextList(288, 128, 8, 40);

	// Set palette
	setInterface("storesInfo");

	add(_window, "window", "storesInfo");
	add(_btnOk, "button", "storesInfo");
	add(_txtTitle, "text", "storesInfo");
	add(_txtItem, "text", "storesInfo");
	add(_txtQuantity, "text", "storesInfo");
	add(_txtSpaceUsed, "text", "storesInfo");
	add(_lstStores, "list", "storesInfo");

	centerAllSurfaces();

	// Set up objects
	_window->setBackground(_game->getMod()->getSurface("BACK13.SCR"));

	_btnOk->setText(tr("STR_OK"));
	_btnOk->onMouseClick((ActionHandler)&StoresState::btnOkClick);
	_btnOk->onKeyboardPress((ActionHandler)&StoresState::btnOkClick, Options::keyOk);
	_btnOk->onKeyboardPress((ActionHandler)&StoresState::btnOkClick, Options::keyCancel);

	_txtTitle->setBig();
	_txtTitle->setAlign(ALIGN_CENTER);
	_txtTitle->setText(tr("STR_STORES"));

	_txtItem->setText(tr("STR_ITEM"));

	_txtQuantity->setAlign(ALIGN_RIGHT);
	_txtQuantity->setText(tr("STR_QUANTITY_UC"));

	_txtSpaceUsed->setAlign(ALIGN_RIGHT);
	_txtSpaceUsed->setText(tr("STR_SPACE_USED_UC"));

	_lstStores->setColumns(3, 162, 57, 57);
	_lstStores->setAlign(ALIGN_RIGHT, 1);
	_lstStores->setAlign(ALIGN_RIGHT, 2);
	_lstStores->setSelectable(true);
	_lstStores->setBackground(_window);
	_lstStores->setMargin(2);

	const std::vector<std::string> &items = _game->getMod()->getItemsList();
	std::vector<StoredItem> storedItems;
	for (std::vector<std::string>::const_iterator item = items.begin(); item != items.end(); ++item)
	{
		int qty = _base->getStorageItems()->getItem(*item);
		if (qty > 0)
		{
			RuleItem *rule = _game->getMod()->getItem(*item, true);
			storedItems.push_back(StoredItem());
			storedItems[storedItems.size() - 1].item = *item;
			storedItems[storedItems.size() - 1].displayName = tr(*item);
			storedItems[storedItems.size() - 1].qty = qty;
			storedItems[storedItems.size() - 1].storageSize = qty * rule->getSize();

		}
	}
	std::sort(storedItems.begin(), storedItems.end(), compareSpace());

	for (std::vector<StoredItem>::const_iterator storedItem = storedItems.begin(); storedItem != storedItems.end(); ++storedItem)
	{
		std::ostringstream quantity, storageSize;
		quantity << (*storedItem).qty;
		storageSize << std::fixed << std::setprecision(1) << (*storedItem).storageSize;
		_lstStores->addRow(3, (*storedItem).displayName.c_str(), quantity.str().c_str(), storageSize.str().c_str());
	}
}

/**
 *
 */
StoresState::~StoresState()
{

}

/**
 * Returns to the previous screen.
 * @param action Pointer to an action.
 */
void StoresState::btnOkClick(Action *)
{
	_game->popState();
}

}

File: DelphiPort/SourceCode/Basescape/StoresState.pas
unit StoresState;

interface

uses
  Classes, SysUtils,
  Engine.State, Engine.Game, Engine.Action,
  Mod.Mod, Engine.LocalizedText, Engine.Options,
  Interface.TextButton, Interface.Window, Interface.Text, Interface.TextList,
  Savegame.Base, Mod.RuleItem, Savegame.ItemContainer;

type
  TStoresState = class(TState)
  private
    FBase: TBase;
    FBtnOk: TTextButton;
    FWindow: TWindow;
    FTxtTitle, FTxtItem, FTxtQuantity, FTxtSpaceUsed: TText;
    FLstStores: TTextList;
  public
    constructor Create(AOwner: TComponent; Base: TBase);
    destructor Destroy; override;
    procedure BtnOkClick(Sender: TObject; Action: TAction);
  end;

implementation

constructor TStoresState.Create(AOwner: TComponent; Base: TBase);
var
  items: TStringList;
  i: Integer;
  rule: TRuleItem;
  qty: Integer;
begin
  inherited Create(AOwner);
  FBase := Base;

  FWindow := TWindow.Create(Self, 320, 200, 0, 0);
  FBtnOk := TTextButton.Create(300, 16, 10, 176);
  FTxtTitle := TText.Create(310, 17, 5, 8);
  FTxtItem := TText.Create(142, 9, 10, 32);
  FTxtQuantity := TText.Create(88, 9, 152, 32);
  FTxtSpaceUsed := TText.Create(74, 9, 240, 32);
  FLstStores := TTextList.Create(288, 128, 8, 40);

  SetInterface('storesInfo');
  Add(FWindow, 'window', 'storesInfo');
  Add(FBtnOk, 'button', 'storesInfo');
  Add(FTxtTitle, 'text', 'storesInfo');
  Add(FTxtItem, 'text', 'storesInfo');
  Add(FTxtQuantity, 'text', 'storesInfo');
  Add(FTxtSpaceUsed, 'text', 'storesInfo');
  Add(FLstStores, 'list', 'storesInfo');
  CenterAllSurfaces;

  FWindow.SetBackground(FGame.GetMod.GetSurface('BACK13.SCR'));
  FBtnOk.SetText(Translate('STR_OK'));
  FBtnOk.OnMouseClick := BtnOkClick;
  FBtnOk.OnKeyboardPress := BtnOkClick;

  FTxtTitle.SetBig;
  FTxtTitle.SetAlign(ALIGN_CENTER);
  FTxtTitle.SetText(Translate('STR_STORES'));
  FTxtItem.SetText(Translate('STR_ITEM'));
  FTxtQuantity.SetText(Translate('STR_QUANTITY_UC'));
  FTxtSpaceUsed.SetText(Translate('STR_SPACE_USED_UC'));

  FLstStores.SetColumns([162, 92, 32]);
  FLstStores.SetSelectable(True);
  FLstStores.SetBackground(FWindow);
  FLstStores.SetMargin(2);

  items := FGame.GetMod.GetItemsList;
  for i := 0 to items.Count - 1 do
  begin
    qty := FBase.GetStorageItems.GetItem(items[i]);
    if qty > 0 then
    begin
      rule := FGame.GetMod.GetItem(items[i], True);
      FLstStores.AddRow([
        Translate(items[i]),
        IntToStr(qty),
        IntToStr(qty * rule.GetSize)
      ]);
    end;
  end;
end;

destructor TStoresState.Destroy;
begin
  inherited;
end;

procedure TStoresState.BtnOkClick(Sender: TObject; Action: TAction);
begin
  FGame.PopState;
end;

end.